subsys/ipc/ipc_service/backends/ipc_icmsg.c

```diff
@@ -0,0 +1,229 @@
+
+
+struct backend_data_t {
+	/* Tx/Rx buffers. */
+	struct icmsg_buf *tx_ib;
+	struct icmsg_buf *rx_ib;
+
+	/* Backend ops for an endpoint. */
+	const struct ipc_service_cb *ops;
+
+	/* General */
+	struct k_work mbox_work;
+	atomic_t state;
+};
+
+struct backend_config_t {
+	uintptr_t tx_shm_addr;
+	uintptr_t rx_shm_addr;
+	size_t tx_shm_size;
+	size_t rx_shm_size;
+	struct mbox_channel mbox_tx;
+	struct mbox_channel mbox_rx;
+};
+
+static void mbox_callback_process(struct k_work *item)
+{
+	struct backend_data_t *dev_data = CONTAINER_OF(item, struct backend_data_t, mbox_work);
+	static uint8_t cb_buffer[CB_BUF_SIZE];
+
+	atomic_t state = atomic_get(&dev_data->state);
+	int len = icmsg_buf_read(dev_data->rx_ib, cb_buffer, CB_BUF_SIZE);
+
+	__ASSERT_NO_MSG(len <= CB_BUF_SIZE);
+
+	if (len == -EAGAIN) {
+		(void)k_work_submit(&dev_data->mbox_work);
+		return;
+	} else if (len <= 0) {
+		return;
+	}
+
+	if (state == ICMSG_STATE_READY) {
+		if (dev_data->ops && dev_data->ops->received) {
+			dev_data->ops->received(cb_buffer, len, NULL);
+		} else {
+			__ASSERT_NO_MSG(false);
+			return;
+		}
+
+		/* Reading 0 to know if there are data in the buffer. */
+		len = icmsg_buf_read(dev_data->rx_ib, NULL, 0);
+		if (len > 0) {
+			(void)k_work_submit(&dev_data->mbox_work);
+		}
+	} else {
+		__ASSERT_NO_MSG(state == ICMSG_STATE_BUSY);
+		if (len == sizeof(magic) && !memcmp(magic, cb_buffer, len)) {
+			if (dev_data->ops && dev_data->ops->bound) {
+				dev_data->ops->bound(NULL);
+			} else {
+				__ASSERT_NO_MSG(false);
+				return;
+			}
+
+			atomic_set(&dev_data->state, ICMSG_STATE_READY);
+		} else {
+			__ASSERT_NO_MSG(false);
+			return;
+		}
+	}
+}
+
+static void mbox_callback(const struct device *instance, uint32_t channel,
+			  void *user_data, struct mbox_msg *msg_data)
+{
+	struct backend_data_t *dev_data = user_data;
+
+	(void)k_work_submit(&dev_data->mbox_work);
+}
+
+static int mbox_init(const struct device *instance)
+{
+	const struct backend_config_t *conf = instance->config;
+	struct backend_data_t *dev_data = instance->data;
+	int err;
+
+	k_work_init(&dev_data->mbox_work, mbox_callback_process);
+
+	err = mbox_register_callback(&conf->mbox_rx, mbox_callback, dev_data);
+	if (err != 0) {
+		return err;
+	}
+
+	return mbox_set_enabled(&conf->mbox_rx, 1);
+}
+
+static int register_ept(const struct device *instance, void **token,
+			const struct ipc_ept_cfg *cfg)
+{
+	const struct backend_config_t *conf = instance->config;
+	struct backend_data_t *dev_data = instance->data;
+	struct ipc_ept *ep = (struct ipc_ept *)token;
+
+	if (atomic_get(&dev_data->state) == ICMSG_STATE_BUSY) {
+		/* Endpoint is still being initialized. */
+		return -EBUSY;
+	}
+
+	if (atomic_get(&dev_data->state) == ICMSG_STATE_READY) {
+		/* Already registered. This backend support single ep. */
+		return -EALREADY;
+	}
+
+	ep->instance = instance;
+	dev_data->ops = &cfg->cb;
+
+	atomic_set(&dev_data->state, ICMSG_STATE_BUSY);
+
+	int ret = mbox_init(instance);
+
```

Comment:
empty line

Reply:
It must be here, because ret is declared here. Zephyr's checkpatch will complain if i remove this line.

Reply:
can you paste the checkpath error? And please move the `ret` definition an the beginning of the function.

Reply:
I can't trigger this locally - i don't know why. Maybe something has change since my last patches to Zephyr. Anyway i will push changes so it will be as you want.